AsyncNetwork implementation as well as unit testing

sphill99 opened the pull request:
This PR defines the AsyncNetwork class, as well as an implementation in the form of BasicAsyncNetwork. All test cases that were covered for BasicNetwork are covered here as unit tests. Some shared code I put into a NetworkUtility class.

Some open questions/comments:
- BasicAsyncNetworkTest and BasicNetwork share one helper method. Wasn't sure if it was worth adding a shared class to not duplicate code.
- Wasn't totally sure if the handling of timeouts before retry is what we're looking for.
- For this to compile, I had to change AsyncHttpStack's setExecutor methods to require(LIBRARY) instead of also including subgroups. I'm looking into this now, but changed it for the time being. Also, since the RequestQueue doesn't see the stack directly, I have methods in the network set the executors in AsyncHttpStack.

sphill99 commented on src/main/java/com/android/volley/toolbox/BasicNetwork.java:
```diff
@@ -136,7 +126,8 @@ public NetworkResponse performRequest(Request<?> request) throws VolleyError {
                                 responseHeaders);
```
I had not considered that, but I'll look into it.

sphill99 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,419 @@
+            return;
+        }
+        request.addMarker(String.format("%s-retry [timeout=%s]", logPrefix, oldTimeout));
+        // If we are using an async stack, then perform retry after a timeout.
```
I had also considered having the network user set the timeout for retrying again. Then we could also check which stack they're using there instead of in this method, and just default to 0 if non-async stack. Wasn't sure though about this.

jpd236 replied:
It looks like what the current code is doing is just trying again after an unconfigurable 100ms delay, which doesn't seem right to me.

What I think we should do for now is match the existing behavior as much as possible with our current APIs. That would mean retrying immediately upon failure. In theory, when the request is retried, getTimeoutMs will return an increased time unit when passed to Cronet. However, as we established earlier, Cronet doesn't actually support setting a timeout, so this increased timeout gets ignored. But that's an implementation detail of the HTTP stack, and other stacks might respect the timeout.

I do think we should make it possible for clients to configure a gap between retries, as the doc noted. But this would be a separate concept which requires a new API, and I don't think we should assume a fixed 100ms retry interval for each retry.

jpd236 commented on src/main/java/com/android/volley/AsyncNetwork.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class AsyncNetwork implements Network {
```
Need Javadocs for any public APIs

jpd236 commented on src/main/java/com/android/volley/AsyncNetwork.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Override
+    public NetworkResponse performRequest(Request<?> request) throws VolleyError {
+        // Is there a better way to go about this?
```
looks mostly fine to me, but personally I'd probably scrap the Response POJO and just have two AtomicReferences, one for NetworkResponse and one for VolleyError. It doesn't seem like the container is buying you much.

sphill99 replied:
Okay, that's what I had, but switched it to keep the same structure as AsyncHttpStack.

jpd236 commented on src/main/java/com/android/volley/toolbox/AsyncHttpStack.java:
```diff
@@ -57,14 +57,14 @@ public abstract void executeRequest(
      * This method must be called before executing any requests.
      */
-    @RestrictTo({RestrictTo.Scope.SUBCLASSES, RestrictTo.Scope.LIBRARY_GROUP})
+    @RestrictTo({RestrictTo.Scope.LIBRARY_GROUP})
```
The problem here is that it prevents someone outside the Volley from extending AsyncHttpStack or AsyncNetwork, which is totally valid. Sounds like you're still looking into why this is happening, but let me know if I can help.

sphill99 replied:
One thing I was unclear on: does implementing these methods directly in AsyncNetwork/AsyncHttpStack solve this issue? Like, would we not need the RestrictTo.Scopt.SUBCLASSES part, since we're doing that?

jpd236 commented on src/main/java/com/android/volley/AsyncNetwork.java:
```diff
@@ -0,0 +1,100 @@
+     * AsyncHttpStack.
+     */
+    @RestrictTo({RestrictTo.Scope.LIBRARY_GROUP})
+    public abstract void setNonBlockingExecutorForStack(ExecutorService executor);
```
WDYT about implementing these methods in AsyncNetwork directly rather than requiring subclasses to do it? E.g. having a protected constructor here which takes the AsyncHttpStack and does the right thing? I'm not sure why any subclass would do anything differently.

You can make the executors accessible to subclasses either by making the variables protected here or (preferably) exposing protected getters.

sphill99 replied:
Yeah, I like this a lot. I can't another implementation for this method.

jpd236 commented on src/main/java/com/android/volley/toolbox/NetworkUtility.java:
```diff
@@ -0,0 +1,134 @@
+ * BasicAsyncNetwork}
+ */
+public class NetworkUtility {
+    protected static final boolean DEBUG = VolleyLog.DEBUG;
```
This seems a bit redundant, vs. just referencing VolleyLog.DEBUG directly.

sphill99 replied:
Yeah, I agree. I had just kept that from BasicNetwork, but doesn't really seem to have a purpose.

jpd236 commented on src/main/java/com/android/volley/toolbox/NetworkUtility.java:
```diff
@@ -0,0 +1,134 @@
+     *     release of Volley.
+     */
+    @Deprecated
+    static Map<String, String> convertHeaders(Header[] headers) {
```
This has to remain where it was - per the Javadoc, it was inadvertently exposed as a publicly accessible API, so moving it might break users of the library.

jpd236 commented on src/main/java/com/android/volley/toolbox/NetworkUtility.java:
```diff
@@ -0,0 +1,134 @@
+     * @param entry The cached response.
+     * @return The combined list of headers.
+     */
+    static List<Header> combineHeaders(List<Header> responseHeaders, Cache.Entry entry) {
```
Some of these header utilities might be good fits for https://github.com/google/volley/blob/master/src/main/java/com/android/volley/toolbox/HttpHeaderParser.java, rather than needing a new class

sphill99 replied:
I'm confused by what you mean here?

jpd236 replied:
I'm suggesting that these methods that relate to header manipulation could be added to HttpHeaderParser instead - this class already has a bunch of utilities around headers, so it seems like we might want to try to keep those together.

sphill99 replied:
oh that makes a lot more sense. I thought you were saying this could be done in that class.

jpd236 commented on src/test/java/com/android/volley/toolbox/BasicAsyncNetworkTest.java:
```diff
@@ -0,0 +1,431 @@
+
+    @Mock private Request<String> mMockRequest;
+    @Mock private RetryPolicy mMockRetryPolicy;
+    private ExecutorService executor = Executors.newSingleThreadExecutor();
```
Generally speaking, we want to avoid using real threads in unit tests as we don't have as much control over when they run. See http://robolectric.org/javadoc/4.3/org/robolectric/android/util/concurrent/InlineExecutorService.html, for example, or just use MoreExecutors.directExecutor() if that works?

sphill99 replied:
Currently we only use version 3.4.2 for Roboelectric dependency. To use that, I think we'd need to use a newer version. Would that be fine? 
Using MoreExecutors.directExecutor() won't work since that provides an Executor and not an ExecutorService. Also, we can use RoboExexutorService instead, but I don't think that runs the tasks right away.

sphill99 replied:
MoreExecutors.newDirectExecutorService() seems like a good option, and works.

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,432 @@
+        request.addMarker(String.format("%s-retry [timeout=%s]", logPrefix, oldTimeout));
+        // If we are using an async stack, then perform retry after a timeout.
+        if (mBaseHttpStack instanceof AsyncHttpStack) {
+            mHandler.postDelayed(
```
We shouldn't need a Handler for this; we should be able to use the Executor. But we might need to change it to a ScheduledExecutorService instead if we decide to keep this behavior.

sphill99 replied:
If what we're passing in is a ThreadPoolExecutor, how could we change it to a ScheduledExecutorService?

jpd236 replied:
We'd need to use something different than a ThreadPoolExecutor in that case. ScheduledThreadPoolExecutor should work.

(That said, this may not be a problem for now if we drop the timing constraint, although we may want to keep the API as requiring a ScheduledExecutorService so that we'll have that option in the future).

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,432 @@
+     *     Apache HTTP. This method may be removed in a future release of Volley.
+     */
+    @Deprecated
+    public BasicAsyncNetwork(HttpStack httpStack) {
```
let's use the Builder pattern here too?

We don't need all these deprecated constructors for a new class. Just don't accept HttpStack from BaseAsyncNetwork.

For that matter, do we actually need to support legacy BaseHttpStacks from BasicAsyncNetwork? Or can we just require that (Basic)Network be used instead in that case?

sphill99 replied:
Are you saying our constructor should only take in an AsyncHttpStack? My only worry is then we have to handle both AsyncNetworks and Networks in AsyncRequestQueue. Unless we still require users to pass in an AsyncNetwork into AsyncRequestQueue. Honestly, if you're using a blocking network and blocking stack (and potentially a blocking cache), does it make sense for them to be able to use the AsyncRequestQueue? With the cache is makes sense, since we require a newish api level, but in this case could we just require they use an AsyncNetwork, which requires an AsyncStack?

jpd236 replied:
Yes, I agree with what you're saying - I don't see any reason to support the blocking flows for Network in the Async stack. Cache is indeed needed for older devices as you say, but we have an AsyncNetwork implementation that should be usable by anyone willing to use AsyncRequestQueue (or if you don't like Cronet, you could use OkHttp instead, which would also work fine), and if you're using a Network, that broadly defeats the purpose of using the async stack to begin with, and you should just use the legacy one.

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,432 @@
+                                PoolingByteArrayOutputStream bytes =
+                                        new PoolingByteArrayOutputStream(
+                                                mPool, httpResponse.getContentLength());
+                                byte[] buffer = new byte[1024];
```
Shouldn't this come from the pool?

sphill99 replied:
Yep

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,432 @@
+                // no-content request.
+                responseContents = new byte[0];
+            } else {
+                Runnable run =
```
A lot of this is copied from the BasicNetwork stream copying code, right? Can we share it?

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,432 @@
+                return;
+            }
+        }
+        runAfterBytesReceived(
```
nit: since this only applies to the case that the response bytes are already read, which is checked many lines above, I'd put this up there and short-circuit return in that case, and then do the rest of the handling for the stream reading case.

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,432 @@
+                                    bytes.write(buffer, 0, count);
+                                }
+                                byte[] finalResponseContents = bytes.toByteArray();
+                                runAfterBytesReceived(
```
nit: perhaps "onResponseRead", which would be consistent with "onRequestSucceeded" or "onRequestFailed" naming-wise.

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    /* Method to be called after a failed network request */
+    private void onRequestFailed(
```
any more code sharing we can do here w/ BasicNetwork to go from Exception type to how it gets handled?

sphill99 commented:
I addressed most comments in the last commit. I shared a lot of code between BasicNetwork and BasicAsyncNetwork in NetworkUtility. I also made it so that AsyncNetwork requires an AsyncHttpStack, so now there is no blocking (unless the input stream needs to be read). I also switched back, so that the request is retried immediately right now. 

A few main comments:
1) I stopped using an actual executor in unit tests, but to use InlineExecutorService, I think we'd need to update to a newer Roboelectric version.
2) Currently, there is no way to wait before retrying again. I'm a little confused on how to move forward with this.
3) I shared a lot of code, and it currently obfuscates what's going on a little bit. Wasn't sure if this was the best way to share the code. If so, I'll add descriptive comments to make it more clear what's happening.

sophiahaim commented on src/main/java/com/android/volley/toolbox/BasicNetwork.java:
```diff
@@ -169,70 +137,32 @@ public NetworkResponse performRequest(Request<?> request) throws VolleyError {
-                        throw new ClientError(networkResponse);
-                    } else if (statusCode >= 500 && statusCode <= 599) {
-                        if (request.shouldRetryServerErrors()) {
-                            attemptRetryOnException(
```
Remove the #attemptRetryOnException method from this class since it's no longer used

sophiahaim commented on src/main/java/com/android/volley/toolbox/NetworkUtility.java:
```diff
@@ -0,0 +1,224 @@
+            }
+            ((AsyncNetwork) network).performRequest(request, callback);
+        } else {
+            network.performRequest(request);
```
I don't see this in the original BasicNetwork#attemptRetryOnException method. Why was it added here? (or maybe I'm missing where this is called in the original implementation)

jpd236 replied:
Yeah,  I don't think this quite works. BasicNetwork works today by running the request in a while(true) loop until this method throws; calling performRequest directly is sort of counter to that since it makes the loop redundant.

I think we either do the same thing in the async network (except we need to deal with asynchronicity) or do something like return a boolean here and let the caller inspect the result to determine if a retry is necessary, rather than passing the network in and making a call back out to it.

sphill99 replied:
In every case, either an error should be thrown (/passed through callback) or the request should be retried. Because of this, I just catch the error, call the callback, and return, or perform the request. Also, now BasicNetwork works the same as it originally did (either error is thrown to get out of while loop, or it's looped again).

sophiahaim commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+
+    protected final ByteArrayPool mPool;
+
+    protected ExecutorService mBlockingExecutor;
```
Now that you set this in AsyncNetwork, I don't think it really makes sense to have it as a class variable here. I think you should just get it (#getBlockingExecutor) in each of the methods that use it (and just set a local var)

sophiahaim commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+                    this);
+        } catch (VolleyError volleyError) {
+            callback.onError(volleyError);
+        }
```
You need a callback in the non-VolleyError case

sphill99 replied:
Now that the callback is called outside of the handleException method, I think it's clearer to the reader that it's always being thrown.

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+/** A network performing Volley requests over an {@link HttpStack}. */
+public class BasicAsyncNetwork extends AsyncNetwork {
+
+    protected final ByteArrayPool mPool;
```
private - do subclasses of BasicAsyncNetwork need direct access to the pool? not sure there's a use case there at the moment.

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+        final List<Header> responseHeaders = httpResponse.getHeaders();
+        // Handle cache validation.
+        if (statusCode == HttpURLConnection.HTTP_NOT_MODIFIED) {
+            callback.onSuccess(
```
missing short-circuit return, I think.

jpd236 commented on src/main/java/com/android/volley/toolbox/NetworkUtility.java:
```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    static NetworkResponse getNetworkResponse(
```
This method name is too generic for what it actually does - return a response for the HTTP_NOT_MODIFIED case. Perhaps getNotModifiedNetworkResponse.

jpd236 commented on src/main/java/com/android/volley/toolbox/NetworkUtility.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    static NetworkResponse getNetworkResponse(
+            int statusCode, Request<?> request, long requestStart, List<Header> responseHeaders) {
```
statusCode seems to be unused.

jpd236 replied:
nit: rather than passing in the start time and calculating the duration yourself, I would suggest that the caller be responsible for passing in the duration. Passing in a start time is a more complex API because the time scale isn't intuitively obvious and easier to get wrong - how do you know/enforce it's based on elapsed realtime?

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+            return;
+        }
+
+        final InputStream inputStream = httpResponse.getContent();
```
Maybe add an explanatory comment here - something like:

"The underlying AsyncHttpStack does not support asynchronous reading of the response into a byte array, so we need to submit a blocking task to copy the response from the InputStream instead."

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+                new Runnable() {
+                    @Override
+                    public void run() {
+                        byte[] finalResponseContents = new byte[0];
```
It shouldn't be necessary to initialize this - it's probably because of the missing short circuit below.

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+                                    NetworkUtility.inputStreamToBytes(
+                                            inputStream, httpResponse.getContentLength(), mPool);
+                        } catch (IOException e) {
+                            onRequestFailed(
```
need to return here to avoid calling onResponseRead

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+        mBlockingExecutor = getBlockingExecutor();
+        if (mBlockingExecutor == null) {
+            throw new IllegalStateException(
+                    "mBlockingExecuter should be set before making a request");
```
nit: s/should/must

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+        logSlowRequests(requestLifetime, request, responseContents, statusCode);
+
+        if (statusCode < 200 || statusCode > 299) {
+            onRequestFailed(
```
missing short-circuit return. Is it possible to unit test for this and/or do a deeper look for cases where this is happening?

sphill99 replied:
Are you asking whether it's possible to unit test whether we return correctly? The only thing I can think of is verifying that methods returning are never called. But that relies on us knowing where they all are. I think just making different types of requests in a test app and making sure they don't make two callback calls is the best way to do it.

jpd236 replied:
Right, my hope is that it would be possible to wire up a request with a mock stack and validate that the expected callback is invoked exactly once and that the other callback methods are not invoked at all.

sphill99 replied:
okay, I'll work on that. I think that shouldn't be too hard actually.

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+                        responseHeaders));
+    }
+
+    public static class Builder {
```
Javadoc for this and public methods

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+            return this;
+        }
+
+        public BasicAsyncNetwork build() {
```
throw is mAsyncStack is null?

sphill99 replied:
is it better to throw, or add the @NonNull annotation? I get a warning if I do both.

jpd236 replied:
OK - we haven't used `@NonNull` much before, but I suppose it's best here because it can be caught at compile time, whereas throwing is only caught at runtime. I'm not 100% convinced b/c it depends on Android Studio tooling and other developers might use other tools that don't detect the problem, but it'll crash at runtime anyway.

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+
+    public static class Builder {
+        private static final int DEFAULT_POOL_SIZE = 4096;
+        AsyncHttpStack mAsyncStack;
```
private

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,235 @@
+    public static class Builder {
+        private static final int DEFAULT_POOL_SIZE = 4096;
+        AsyncHttpStack mAsyncStack;
+        ByteArrayPool mPool;
```
private

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicNetwork.java:
```diff
@@ -169,70 +137,32 @@ public NetworkResponse performRequest(Request<?> request) throws VolleyError {
+     *     release of Volley.
+     */
+    @Deprecated
+    static Map<String, String> convertHeaders(Header[] headers) {
```
this was protected before - it should remain protected.

jpd236 commented on src/main/java/com/android/volley/toolbox/NetworkUtility.java:
```diff
@@ -0,0 +1,224 @@
+ * Utility class for methods that are shared between {@link BasicNetwork} and {@link
+ * BasicAsyncNetwork}
+ */
+public class NetworkUtility {
```
mark final, add private constructor to prevent instantiation of this utility class

jpd236 commented on src/test/java/com/android/volley/toolbox/BasicAsyncNetworkTest.java:
```diff
@@ -0,0 +1,435 @@
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(sdk = 16)
```
Is this needed? What for?

sphill99 replied:
That is no longer needed. I had tried something that required a higher API, but didn't use it.

sphill99 commented:
I believe I made all the appropriate fixes from the comments. I no longer pass in the network or callback into handleException. It either throws the exception, or does nothing. If no exception is thrown, the request is retried (by calling performRequest in async version, and by looping again in non-async version). I believe all the short circuit return issues have been addressed. Will comb through again just to make sure. 

One thing to note. I couldn't make the setter methods in AsyncHttpStack protected, since AsyncNetwork (which calls the setters), is not in the same package. Currently they're public, but we could move AsyncHttpStack outside of the toolbox.

sphill99 commented:
Just added testing that makes sure that the callback is only ever called once. Added a mock callback, and then verified the number of times onSuccess and onError are called. I'm not super familiar with Mockito though, so not sure if this is the best way to do this.

sophiahaim commented on src/main/java/com/android/volley/AsyncNetwork.java:
```diff
@@ -0,0 +1,124 @@
+
+    /**
+     * This method sets the non blocking executor to be used by the stack for non-blocking tasks.
+     * This method must be called before performing any requests if you are using an AsyncHttpStack.
```
Remove (since always using an AsyncHttpsStack now)

sophiahaim commented on src/main/java/com/android/volley/cronet/CronetHttpStack.java:
```diff
@@ -70,7 +64,7 @@ public void executeRequest(
             final OnRequestComplete callback) {
-        if (mBlockingExecutor == null || mCallbackExecutor == null) {
+        if (getBlockingExecutor() == null || getNonBlockingExecutor() == null) {
             throw new IllegalStateException("Must set blocking / non-blocking executors");
```
super nit, but I think this should be an "and" ("Must set blocking and non-blocking executors")

sphill99 replied:
Shouldn't it be that if either of them are null, we throw the exception?

sophiahaim replied:
Yep, I'm just referring to the IllegalStateException method, since both should be set. But yes, the code itself is right (should throw if either are null since both need to be set)

sophiahaim replied:
*the IllegalStateException message

jpd236 commented on src/main/java/com/android/volley/AsyncNetwork.java:
```diff
@@ -0,0 +1,124 @@
+        mAsyncStack = stack;
+    }
+
+    public interface OnRequestComplete {
```
javadoc

jpd236 commented on src/main/java/com/android/volley/AsyncNetwork.java:
```diff
@@ -0,0 +1,124 @@
+        mAsyncStack.setBlockingExecutor(executor);
+    }
+
+    protected ExecutorService getBlockingExecutor() {
```
javadoc - in particular, what sorts of tasks should subclasses use this executor for?

also for getNonBlockingExecutor and getHttpStack.

jpd236 commented on src/main/java/com/android/volley/AsyncNetwork.java:
```diff
@@ -0,0 +1,124 @@
+     * This method must be called before performing any requests.
+     */
+    @RestrictTo({RestrictTo.Scope.LIBRARY_GROUP})
+    public void setNonBlockingExecutorForStack(ExecutorService executor) {
```
"setNonBlockingExecutor"? Not sure what the "ForStack" is for here since it's no different from what setBlockingExecutor is doing. This executor could be used by the network too.

sphill99 replied:
My reasoning was that getNonBlockingExecutor() isn't called at all by BasicAsyncNetwork. But you're right that it could be used by the network.

jpd236 commented on src/main/java/com/android/volley/AsyncNetwork.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    /**
+     * This method sets the non blocking executor to be used by the stack for non-blocking tasks.
```
by the network and stack?

jpd236 commented on src/main/java/com/android/volley/toolbox/AsyncHttpStack.java:
```diff
@@ -57,15 +59,27 @@ public abstract void executeRequest(
+        mBlockingExecutor = executor;
+    }
+
+    protected ExecutorService getBlockingExecutor() {
```
javadoc for this and getNonBlockingExecutor

sophiahaim replied:
I think this javadoc wasn't included in your last commit, please add, thanks!

jpd236 commented on src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java:
```diff
@@ -0,0 +1,244 @@
+                        } catch (ServerError serverError) {
+                            // This should never happen since we already check if inputStream is
+                            // null
+                            throw new RuntimeException(serverError);
```
perhaps IllegalStateException - or consider changing inputStreamToBytes not to throw ServerError and moving that logic into BasicNetwork if it's unused in BasicAsyncNetwork's flow.

sphill99 replied:
I think handling it in BasicNetwork is better, and then not dealing with the ServerError in BasicAsyncNetwork.

jpd236 commented on src/test/java/com/android/volley/toolbox/BasicAsyncNetworkTest.java:
```diff
@@ -0,0 +1,509 @@
+@RunWith(RobolectricTestRunner.class)
+public class BasicAsyncNetworkTest {
+
+    CompletableFuture<NetworkResponse> future = new CompletableFuture<>();
```
nit: private

jpd236 commented on src/test/java/com/android/volley/toolbox/BasicAsyncNetworkTest.java:
```diff
@@ -0,0 +1,509 @@
+
+    CompletableFuture<NetworkResponse> future = new CompletableFuture<>();
+    @Mock private RetryPolicy mMockRetryPolicy;
+    @Mock AsyncNetwork.OnRequestComplete mockCallback;
```
private

jpd236 commented on src/test/java/com/android/volley/toolbox/BasicAsyncNetworkTest.java:
```diff
@@ -0,0 +1,509 @@
+        reset(mMockRetryPolicy, mockCallback);
+    }
+
+    @Test(expected = IllegalStateException.class)
```
Can we use Assert.assertThrows without adding a new dependency? If so, that's preferable since it more tightly describes the code expected to fail, rather than just letting any IllegalStateException thrown anywhere inside this test cause the test to pass.

sphill99 replied:
I don't think so, since that's only in JUnit5. I'm not aware of another way to test that that specific code causes it.

sphill99 commented:
I believe I've now addressed all comments. The only thing I didn't change was using assertThrows, but I don't believe we are able to since it requires JUnit 5.
